Objects/codeobject.c

```diff
@@ -12,6 +12,66 @@
 #include "clinic/codeobject.c.h"
 
 
+static void
+notify_code_watchers(PyCodeEvent event, PyCodeObject *co)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    if (interp->active_code_watchers) {
```

Comment:
I'm OK with this as is, but you could speed this up with an invariant.
`(active_code_watchers & (1 << i))` implies `interp->code_watchers[i] != NULL`

Then the loop can be something like this:
```
uint8_t watchers = interp->active_code_watchers;
while (watchers) {
    int watcher = left_most_bit(watchers);
    watchers &= ~(1 << watcher);
    if (interp->code_watchers[watcher](...) {
          PyErr_WriteUnraisable(...);
    }
}
```

This probably applies to the other watchers as well, so you might want to do this in another PR.

Reply:
I think this optimization applies to func watchers, but is already effectively applied for dict and type watchers; in those cases we already never look up the watcher unless the bit is set.

Reply:
I have some more watchers-wide cleanup I wanted to do, so will add this optimization to the next PR, thanks!

Reply:
cleanup and polish PR (including this optimization) is up at GH-99998